### PR TITLE
MinHeap optimization

### PR DESCRIPTION
**Pull request by stephenwithav**
Improve MinHeap.Parent, along with eliminating the need for temporary variables in swapping.

### CODE REVIEW

File: backend/min_heap.go [hunk 1]
```diff
@@ -114,24 +113,23 @@ func (h *MinHeap) Size() int {
...
+	i := h.Size()
+
+	h.Store = append(h.Store, item)
+
+	// Re-sort slice to put the new item in the proper place.
 	for i > 0 && h.Store[h.Parent(i)].Key() > h.Store[i].Key() {
-		temp := h.Store[i]
-		h.Store[i] = h.Store[h.Parent(i)]
-		h.Store[h.Parent(i)] = temp
+		// Swap item locationss.
```
Comment (vimukthi-git) [inline backend/min_heap.go]: 👍

File: backend/min_heap.go [hunk 2]
```diff
@@ -114,24 +113,23 @@ func (h *MinHeap) Size() int {
...
+		h.Store[i], h.Store[h.Parent(i)] = h.Store[h.Parent(i)], h.Store[i]
+
 		i = h.Parent(i)
 	}
 }
 
 func (h *MinHeap) Parent(i int) int {
-	return int(math.Floor(float64(i / 2)))
+	return i >> 1
```
Comment (vimukthi-git) [inline backend/min_heap.go]: 👍

File: backend/min_heap.go [hunk 3]
```diff
@@ -168,18 +166,16 @@ func (h *MinHeap) MinHeapify(i int) {
 	// log.Println("smallest=", smallest)
 	if smallest != i {
 		// log.Println("smallest=", smallest, ", i=", i)
-		temp := h.Store[i]
-		h.Store[i] = h.Store[smallest]
-		h.Store[smallest] = temp
+		h.Store[i], h.Store[smallest] = h.Store[smallest], h.Store[i]
```
Comment (vimukthi-git) [inline backend/min_heap.go]: 👍

File: backend/min_heap.go [hunk 4]
```diff
@@ -114,24 +113,23 @@ func (h *MinHeap) Size() int {
...
-	// log.Println("queue", h, i)
-	if item.Key() > h.Store[i].Key() {
-		log.Error(h, h.Store[i], item)
-		log.Fatal("new key can not be larger than the current")
-	}
-	h.Store[i] = item
-	//log.Println(h.Size, key)
+func (h *MinHeap) DecreaseKey(item architecture.PriorityQueueItem) {
+	// Index of next slot in slice.
+	i := h.Size()
```
Comment (vimukthi-git) [inline backend/min_heap.go]: Nice, This is something that evolved, so didn't realise the simpler way :D

Comment (stephenwithav) [inline backend/min_heap.go]: Totally understandable. :)

Feel free to merge this when you're ready.

**Comment by vimukthi-git** [comment]
@stephenwithav Good thanks for the update. Since this is a change in the main store, I will test your change locally and then merge.

**Comment by stephenwithav** [comment]
@vimukthi-git Have you been able to test this?

**Comment by stephenwithav** [comment]
Did some more tidying up.  The tests pass.

**Comment by vimukthi-git** [comment]
I will merge this in the weekend as there are many edge cases I want to check with this change.

**Comment by stephenwithav** [comment]
I'm sorry to hear that, @vimukthi-git.  I hope you feel better!

**Comment by vimukthi-git** [comment]
Thanks @stephenwithav
